src/java/org/apache/cassandra/service/reads/range/EndpointGroupingCoordinator.java

```diff
@@ -0,0 +1,338 @@
+    private final DataLimits.Counter counter;
+    private final Map<InetAddressAndPort, EndpointQueryContext> endpointContexts;
+    private final List<ReadCallback<EndpointsForRange, ReplicaPlan.ForRangeRead>> perRangeHandlers;
+    private final List<PartitionIterator> concurrentQueries;
+
+    private final long queryStartNanoTime;
+    private final int vnodeRanges;
+
+    /**
+     * @param command current range read command
+     * @param counter the unlimited counter for the command
+     * @param replicaPlans to be queried
+     * @param concurrencyFactor number of vnode ranges to query at once
+     * @param queryStartNanoTime the start time of the query
+     */
+    public EndpointGroupingCoordinator(PartitionRangeReadCommand command,
+                                       DataLimits.Counter counter,
+                                       Iterator<ReplicaPlan.ForRangeRead> replicaPlans,
+                                       int concurrencyFactor,
+                                       long queryStartNanoTime)
+    {
+        this.command = command;
+        this.counter = counter;
+        this.queryStartNanoTime = queryStartNanoTime;
+        this.endpointContexts = new HashMap<>();
+
+        // Read callbacks in token order
+        perRangeHandlers = new ArrayList<>(concurrencyFactor);
+        // Range responses in token order
+        concurrentQueries = new ArrayList<>(concurrencyFactor);
+        int vnodeRanges = 0;
+
+        while (replicaPlans.hasNext() && vnodeRanges < concurrencyFactor)
+        {
+            ReplicaPlan.ForRangeRead replicaPlan = replicaPlans.next();
+
+            boolean isFirst = vnodeRanges == 0;
+            vnodeRanges += replicaPlan.vnodeCount();
+            concurrentQueries.add(createResponse(replicaPlan, isFirst));
+        }
+        this.vnodeRanges = vnodeRanges;
+    }
+
+    public int vnodeRanges()
+    {
+        return vnodeRanges;
+    }
+
+    public PartitionIterator execute()
+    {
+        for (EndpointQueryContext replica : replicas())
+            replica.queryReplica();
+
+        return counter.applyTo(PartitionIterators.concat(concurrentQueries));
+    }
+
+    @VisibleForTesting
+    Collection<EndpointQueryContext> endpointRanges()
+    {
+        return endpointContexts.values();
+    }
+
+    /**
+     * @return number of endpoints to be queried
+     */
+    int endpoints()
+    {
+        return endpointContexts.size();
+    }
+
+    private Collection<EndpointQueryContext> replicas()
+    {
+        return endpointContexts.values();
+    }
+
+    /**
+     * Create a {@link SingleRangeResponse} for a given vnode range. The responses are collected and concatenated by
+     * {@code execute}.
+     */
+    private SingleRangeResponse createResponse(ReplicaPlan.ForRangeRead replicaPlan, boolean isFirst)
+    {
+        PartitionRangeReadCommand subrangeCommand = command.forSubRange(replicaPlan.range(), isFirst);
+        replicaPlan.candidates().size();
```

Comment:
nit: is it needed?

Reply:
No it isn't. Thanks, I've no idea why that was there. I'll remove before merging.